control-plane/agents/src/bin/core/controller/reconciler/volume/garbage_collector.rs

```diff
@@ -250,6 +258,53 @@ async fn disown_unused_replicas(
     GarbageCollector::squash_results(results)
 }
 
+/// Volume replicas could have been possibly expanded as part of volume resize,
+/// but the volume expansion failed overall leaving behind expanded replicas. So we
+/// bring back the replica size to what is expected by the volume spec.
+#[tracing::instrument(level = "debug", skip(context, volume), fields(volume.uuid = %volume.uuid(), request.reconcile = true))]
+async fn revert_expanded_replicas(
+    volume: &mut OperationGuardArc<VolumeSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let target = context.specs().volume_target_nexus_rsc(volume.as_ref());
+    if let Some(tgt) = target {
+        let nexus_size = tgt.immutable_ref().size;
+        debug_assert_eq!(nexus_size, volume.as_ref().size);
+        if nexus_size != volume.as_ref().size {
+            // This should never happen, but just ensuring we don't touch replicas of such a volume.
+            volume.warn_span(|| tracing::info!(volume_size = %volume.as_ref().size, nexus_size = %nexus_size, "Nexus and Volume size mismatch!"));
+            return PollResult::Ok(PollerState::Idle);
+        }
+    }
+
+    for replica in context.specs().volume_replicas(volume.uuid()) {
+        let mut replica = match replica.operation_guard() {
+            Ok(guard) => guard,
+            Err(_) => continue,
+        };
+
+        if (replica.as_ref().size != volume.as_ref().size) && volume.as_ref().operation.is_none() {
```

Comment:
Isn't this check done up the call stack? volume.as_ref().operation.is_none()

```suggestion
        if (replica.as_ref().size != volume.as_ref().size) {
```

Reply:
Yea I guess it's redundant to check here because if there was a pending operation we won't be holding the volume guard here.